brmp/__init__.py

```diff
@@ -48,146 +105,77 @@ def defm(formula_str, df, family=None, priors=None, contrasts=None):
                       encoding.
     :type contrasts: dict
     :return: A wrapper around the model description and the design matrices.
-    :rtype: brmp.DefmResult
+    :rtype: brmp.ModelAndData
 
     Example::
 
       df = pd.DataFrame({'y': [1., 2.], 'x': [.5, 0.]})
-      model = defm('y ~ 1 + x', df)
-
+      model = brm('y ~ 1 + x', df)
 
     """
     assert type(formula_str) == str
     assert type(df) == pd.DataFrame
     assert family is None or type(family) == Family
     assert priors is None or type(priors) == list
     assert contrasts is None or type(contrasts) == dict
-
-    family = family or Normal
-    priors = priors or []
-    contrasts = contrasts or {}
-
-    # TODO: Consider accepting nested arrays as well as numpy arrays.
-    # (If we do, convert to numpy arrays here in `defm`?)
-    assert all(type(val) == np.ndarray and len(val.shape) == 2 for val in contrasts.values())
-
-    formula = parse(formula_str)
-    # Perhaps design matrices ought to always have metadata (i.e.
-    # column names) associated with them, as in Patsy.
     metadata = metadata_from_df(df)
-    desc = makedesc(formula, metadata, family, priors, code_lengths(contrasts))
-    data = makedata(formula, df, metadata, contrasts)
-    return DefmResult(formula, metadata, contrasts, desc, data)
+    define_model_result = define_model(formula_str, metadata, family, priors, contrasts)
+    data = define_model_result.encode(df)
+    return ModelAndData(define_model_result, df, data)
 
 
-# A wrapper around a pair of model and data. Has a friendly `repr` and
-# makes it easy to fit the model.
-class DefmResult:
-    def __init__(self, formula, metadata, contrasts, desc, data):
-        self.formula = formula
-        self.metadata = metadata
-        self.contrasts = contrasts
-        self.desc = desc
+class ModelAndData:
+    def __init__(self, define_model_result, df, data):
+        self.define_model_result = define_model_result
+        self.df = df
         # TODO: Turn this into a `@property` to improve generate docs?
         self.data = data
         """
         A dictionary with the encoded data as its values.
         Each value is a :class:`~numpy.ndarray`.
         """
 
-    def fit(self, backend=_default_backend, algo='nuts', **kwargs):
-        """
-        Fits the wrapped model.
-
-        :param backend: The backend used to perform inference.
-        :type backend: brmp.backend.Backend
-        :param algo: The algorithm used for inference, either ``'nuts'``, ``'svi'`` or ``'prior'``.
-        :type algo: str
-        :param kwargs: Inference algorithm-specific keyword arguments. See the methods on
-                       :class:`~brmp.GenerateResult` for details.
-        :return: A model fit.
-        :rtype: brmp.fit.Fit
-
-        Example::
-
-          fit = defm('y ~ x', df).fit()
+        # We might eventually want to cache generated assets. This
```

Comment:
I think this will be a big benefit specially for numpyro where compilation can take several seconds.